### PR TITLE
Internal features plumbing and new dialog! 

### PR DESCRIPTION
**Pull request by Bonkles**
This PR adds two new things: 
1) Build time YAML configuration to adjust rapid's runtime behavior. 
2) A new feature dependent on this configuration; namely, a new 'Internal features' dialog that allows our internal mappers to enable certain functionality. 

Here's the new dialog: 
interna-features-dialog

There are three options in this dialog, all are checkboxes: 

1) AI Feature Halo
This feature causes an ML pink halo around any features that started life as a Rapid AI feature. This is to say that if a person added an ML road a few months ago, with this option on the road will appear to have a pink outline. 

2) Tag-agnostic Road Combining
By default, two roads with different tags cannot be combined. With this option checked, selecting two end-to-end roads and hitting the 'Combine' key will overwrite the second selected road with the tags from the first. 

3) Set all roads to source=digitalglobe
This treats any roads drawn by the mapper as though they were roads created by the Machine Learning service.

### CODE REVIEW

File: build.js [hunk 1]
```diff
@@ -12,10 +12,14 @@ buildAll();
 function buildAll() {
   if (_currBuild) return _currBuild;
 
+  var build_type = process.env.BUILD_TYPE || 'public'; 
+
   return _currBuild =
+
```
Comment (Bonkles) [inline build.js]: Strike this empty space!

File: build_data.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ const request = require('request').defaults({ maxSockets: 1 });
 let _currBuild = null;
 
 
-function buildData() {
+function buildData(build_type) {
```
Comment (Bonkles) [inline build_data.js]: the 'BUILD_TYPE' environment variable can be overridden in the package.json file in one of the new 'internal' build rules.

File: build_data.js [hunk 3]
```diff
@@ -771,14 +772,14 @@ function translationsToYAML(translations) {
...
+  var readRapidConfig = readFileProm('data/rapid_config.yaml', 'utf8'); 
 
-    return Promise.all([readRapidConfig]).then(function(data) {
-        var config = YAML.load(data[0]); 
+  return Promise.all([readRapidConfig]).then(function(data) {
+      var config = YAML.load(data[0]); 
 
-        return writeFileProm('data/rapid_config.json', JSON.stringify(config, null, 4)); 
-    }); 
+      return writeFileProm('data/rapid_config.json', JSON.stringify(config[build_type], null, 4));
```
Comment (Bonkles) [inline build_data.js]: Depending on the build type, we grab a different config block to place into the config.json. This means that the rapid runtime code does not need to decide which config block to evaluate- it only ever has the one that we switch in & out at build time.

File: data/rapid_config.json [hunk 4]
```diff
@@ -1,7 +0,0 @@
-{
```
Comment (Bonkles) [inline data/rapid_config.json]: This file has been removed from source control, but after an 'npm build' or 'npm build-internal' it would be a jsonified version of whatever is inside data/rapid_config.yaml's "public" or "internal" data block, respectively.

File: data/rapid_config.yaml [hunk 5]
```diff
@@ -2,8 +2,13 @@
 # Rapid configuration-  for customizing runtime behavior. 
 # The use case for this file is to enable/disable features without
 # needing to open/edit/rebuild the code. 
-rapid_config: 
-  # This setting styles any features in the map with a special shadow effect if 
-  # and only if they were originally presented to the user as AI features
-  style_fb_ai_roads:
-    enabled: false
+internal:
+  rapid_config:
```
Comment (Bonkles) [inline data/rapid_config.yaml]: This 'rapid_config' block is selected if we run an 'npm build-internal'. The other public one is selected if we just do an 'npm build'.

File: data/shortcuts.json [hunk 6]
```diff
@@ -85,9 +81,8 @@
               "text": "shortcuts.browsing.display_options.minimap"
             },
             {
-              "modifiers": ["⇧"],
-              "shortcuts": ["map_data.highlight_way_edits.key"],
-              "text": "shortcuts.browsing.display_options.highlight_way_edits",
+              "shortcuts": ["map_data.highlight_edits.key"],
```
Comment (Bonkles) [inline data/shortcuts.json]: This change was not strictly a thing I fixed for this functionality- it was a mistake that I munged while merging from master. This puts it back where it belongs.

File: modules/ui/rapid_internal_features_storage.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
+var memStorage = {}; 
+
+
+export function rapidInternalFeaturesStorage() {
+    // Cloned from context, but we need storage from within the actions
```
Comment (Bonkles) [inline modules/ui/rapid_internal_features_storage.js]: This means that window.localStorage will have some fairly obvious keys in it that a user could enable/disable if they wanted to. Is this bad? Should these be obfuscated more?

File: modules/ui/tools/rapid_internal_features.js [hunk 8]
```diff
@@ -1,42 +1,36 @@
 import _debounce from 'lodash-es/debounce';
```
Comment (Bonkles) [inline modules/ui/tools/rapid_internal_features.js]: Don't worry about the file rename here- I simply noticed that the now-defunct 'fb_roads_toggle' tool from v2 never got pruned from the v3 code. So, I repurposed it since we needed a new tool and renamed it.

File: modules/ui/rapid_internal_features_dialog.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            .classed('disabled', options.greyout)
+            .attr('id', 'section-' + options.id)
+
+        var toggleOptionContainer =  toggleOption.append('div')
+            .attr('class', 'rapid-feature-label-container internal'); 
+
+        toggleOptionContainer
+            .append('div')
+            .attr('id', options.imgid)
+            .attr('class', 'rapid internal feature animation');
```
Comment (chrisklaiber) [inline modules/ui/rapid_internal_features_dialog.js]: That's a lot of generic words as CSS class names! Can we invert this? Like rapid-internal at the top-level or something? And then we can use `feature` after that, if it's a normal iD way of doing these things. 

We can sync on this. Generally spreading generic names around as CSS classes will just bite you later when  you have conflicts. Specific names are easier to target and maintain in CSS's global namespace.

File: modules/ui/rapid_internal_features_storage.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    // modules which are context-free.
```
Comment (chrisklaiber) [inline modules/ui/rapid_internal_features_storage.js]: Ah, I was hoping to be able to make this part of `rapid_context.js`, but it makes sense that these values are "context-free" and done at build-time.

File: data/rapid_config.yaml (hunk 5 above)
Comment (chrisklaiber) [reply data/rapid_config.yaml]: What about adding default values for the various features in here? Right now they all appear to  default to `true` or `false`, perhaps, but you don't know  what  available as a developer without hunting through the code. It might be handy to list them here as initial values, and it's also lets us choose which features to have on by default.

Comment (Bonkles) [reply data/rapid_config.yaml]: Ah, you mean default values for features, orthogonal to whether the 'internal' or 'public' block is selected?

File: css/80_app_fb.css [hunk 11]
```diff
@@ -117,14 +128,29 @@ button.ai-features-toggle.layer-off use {
...
+.modal-rapid .fillL.rapid-stack > .modal-section.rapid-checkbox:first-of-type {
+    margin-top: -10px; 
+    padding-bottom: 0px; 
+    padding-top: 5px; 
+} 
+
+.modal-rapid .fillL.rapid-stack > .modal-section.rapid-checkbox:not(:first-of-type) {
+    padding-top: 2px;
+    padding-bottom: 2px; 
+}
```
[the remark this answers is not in the document]
Comment (Bonkles) [reply css/80_app_fb.css]: This is one of those 'Why did I write it like that' moments. Thanks for catching! Yes, the cascade is our friend....

[the remark this answers is not in the document]
Comment (Bonkles) [reply css/80_app_fb.css]: Additionally, I wonder if the css preprocessor is smart enough to realize when 'if this/if not this' directives are happening and reshuffle them to be more of the form you suggest, with 'default/non-default' clauses.

File: modules/ui/rapid_internal_features_dialog.js (hunk 9 above)
Comment (Bonkles) [reply modules/ui/rapid_internal_features_dialog.js]: The reason for the code discrepancy between this feature vs. the other two is the other two only manifest their behavior after the user does something- i.e. a specific interaction with iD. 

This feature affects the runtime of iD insofar as it changes the way things are styled on the map. this was my way of wiring things up so that the setting would be read initially before the first paint. That said, looking at this code it is jarring to see one setting getting what appears to be 'preferential' treatment from the others.

File: modules/ui/rapid_internal_features_dialog.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+export function uiRapidInternalFeaturesDialog(context) {
+    var featureEnabled = rapidInternalFeaturesStorage(); 
+    var aiFeatureHaloEnabled = featureEnabled('aiFeatureHalo');
+
+
+    function toggleAiFeatureHalo () {
+        var osmLayer = d3_select('.data-layer.osm'); 
+
+        aiFeatureHaloEnabled = !aiFeatureHaloEnabled;
```
[the remark this answers is not in the document]
Comment (Bonkles) [reply modules/ui/rapid_internal_features_dialog.js]: Good catch, I think the difference in code is vestigial. I originally did the initial read of the settings as well as the toggling of the settings from this one file- I thought better of it and split them into separate files. The original bug (and the reason for this closure reference) was to read the setting before we painted the map upon startup of RapiD- otherwise, the setting wouldn't get read until the user had opened the settings dialog. *Not* ideal, hence the difference. 

TL;DR: it shouldn't have to be like this. I will feex.

File: modules/ui/rapid_internal_features_dialog.js (hunk 9 above)
Comment (Bonkles) [reply modules/ui/rapid_internal_features_dialog.js]: Fixed to use 'rapid-feature-animation' here.

**Comment by Bonkles** [comment]
> Maybe we can coin a new term for it? Like "Power" or "Advanced" or "Extra" or maybe we just call them feature flags or gated features. All these imply something beyond normal operation.

I have replaced 'internal' with 'poweruser' everywhere in the yaml, js, and css.  There is one exception, I kept the 'internal' moniker in the package.json and the build.js files. I think this puts some better semantic meaning in the rapid_config.yaml: 

`internal:
  rapid_config:
    poweruser_features_dialog:
      enabled: true

public:
  rapid_config:
    poweruser_features_dialog:
      enabled: true`

File: modules/actions/join.js [hunk 13]
```diff
@@ -71,6 +71,13 @@ export function actionJoin(ids) {
             graph = actionDeleteWay(way.id)(graph);
         });
 
+        if (rapidInternalFeaturesStorage()('tagnosticRoadCombine')){
```
[the remark this answers is not in the document]
Comment (Bonkles) [reply modules/actions/join.js]: Instead of exporting the raw method for 'featureEnabled' and immediately invoking it, I export an object with the 'featureEnabled' method on it. This changes the invocation of the module to something more akin to what we see elsewhere in the code: 
rapidInternalFeatureStorage().featureEnabled('blah');

**Comment by Bonkles** [comment]
Also, the current build problems with the 'tag info ' tests are because of an issue with the verison of iD master we're based off of.  I'll rebase off the latest and retry.

**Comment by Bonkles** [comment]
Hm, seems the buidl problem with the tests still exists even after rebasing to the very latest version of iD master. Will consult with Quincy on this- I'm sure it'll get resolved while they tease apart the new v3 version branch.

**Comment by chrisklaiber** [comment]
Looks good to me once you're feeling fine about the merge, build, checks, etc.

**Comment by Bonkles** [comment]
Squarshed to remove history, incorporate latest changes from iD v3.
